src/pipeline.rs

```diff
@@ -17,30 +35,131 @@ impl Sink for Pipeline {
         &mut self,
         item: Self::SinkItem,
     ) -> Result<AsyncSink<Self::SinkItem>, Self::SinkError> {
-        self.inner.start_send(item)
+        match self.try_flush() {
+            Ok(Async::NotReady) => Ok(AsyncSink::NotReady(item)),
+            Ok(Async::Ready(())) => {
+                // Note how this gets **swapped** with `new_working_set` in the loop.
+                // At the end of the loop, it will only contain finalized events.
+                let mut working_set = vec![item];
+                for inline in self.inlines.iter_mut() {
+                    let mut new_working_set = Vec::with_capacity(working_set.len());
+                    for event in working_set.drain(..) {
+                        inline.transform(&mut new_working_set, event);
+                    }
+                    core::mem::swap(&mut new_working_set, &mut working_set);
+                }
+                self.enqueued.extend(working_set);
+                Ok(AsyncSink::Ready)
+            }
+            Err(e) => Err(e),
+        }
     }
 
     fn poll_complete(&mut self) -> Poll<(), Self::SinkError> {
+        futures01::try_ready!(self.try_flush());
+        debug_assert!(self.enqueued.is_empty());
         self.inner.poll_complete()
     }
 }
 
 impl Pipeline {
     #[cfg(test)]
-    pub fn new_test() -> (Self, Receiver<Event>) {
-        Self::new_with_buffer(100)
+    pub fn new_test(inlines: Vec<Box<dyn FunctionTransform>>) -> (Self, Receiver<Event>) {
```

Comment:
https://github.com/timberio/vector/pull/4941#discussion_r520944681

> I might prefer we leave this method as-is and add another helper if/when we need it to include inlines. That way we don't need to change everywhere we use this without setting any.

I think this is fine, I'll add a builder.